source/opt/struct_packing_pass.cpp

```diff
@@ -0,0 +1,463 @@
+  // Array stride is equal to aligned size of element type
+  const uint32_t elementSize = getPackedSize(*arrayType.element_type());
+  const uint32_t alignment = getPackedAlignment(arrayType);
+  return alignPow2(elementSize, alignment);
+}
+
+uint32_t StructPackingPass::getArrayLength(
+    const analysis::Array& arrayType) const {
+  return getConstantInt(arrayType.LengthId());
+}
+
+uint32_t StructPackingPass::getConstantInt(spv::Id id) const {
+  auto it = constantsMap_.find(id);
+  assert(it != constantsMap_.end() &&
+         "Failed to map SPIR-V instruction ID to constant value");
+  const analysis::Type* constType =
+      context()->get_type_mgr()->GetType(it->second->type_id());
+  assert(constType != nullptr &&
+         "Failed to map SPIR-V instruction result type to definition");
+  assert(constType->kind() == analysis::Type::kInteger &&
+         "Failed to map SPIR-V instruction result type to integer type");
+  return it->second->GetOperand(2).words[0];
+}
+
+StructPackingPass::PackingRules StructPackingPass::ParsePackingRuleFromString(
+    const std::string& s) {
+  if (s == "std140") return PackingRules::Std140;
+  if (s == "std140EnhancedLayout") return PackingRules::Std140EnhancedLayout;
+  if (s == "std430") return PackingRules::Std430;
+  if (s == "std430EnhancedLayout") return PackingRules::Std430EnhancedLayout;
+  if (s == "hlslCbuffer") return PackingRules::HlslCbuffer;
+  if (s == "hlslCbufferPackOffset") return PackingRules::HlslCbufferPackOffset;
+  if (s == "scalar") return PackingRules::Scalar;
+  if (s == "scalarEnhancedLayout") return PackingRules::ScalarEnhancedLayout;
+  return PackingRules::Undefined;
+}
+
+StructPackingPass::StructPackingPass(const char* structToPack,
+                                     PackingRules rules)
+    : structToPack_{structToPack != nullptr ? structToPack : ""},
+      packingRules_{rules} {}
+
+Pass::Status StructPackingPass::Process() {
+  if (packingRules_ == PackingRules::Undefined) return Status::Failure;
+
+  buildConstantsMap();
+
+  // Find all structs of interest
+  uint32_t structIdToPack = ~0u;
+  for (Instruction& instr : context()->module()->debugs2()) {
+    if (instr.opcode() == spv::Op::OpName &&
+        instr.GetOperand(1).AsString() == structToPack_) {
+      structIdToPack = instr.GetOperand(0).AsId();
+      break;
+    }
+  }
+
+  struct StructMemberInfo {
+    spv::Id typeId = 0;
+    analysis::Type* type = nullptr;
+    uint32_t offset = 0; // Original offset
+    uint32_t packedOffset = 0; // Packed offset
+  };
+  std::vector<StructMemberInfo> structMembers;
+
+  if (structIdToPack == ~0u) {
+    return Status::SuccessWithoutChange;
+  }
+
+  bool modified = false;
+
+  // Find all struct member types
+  for (auto
+           it =
+               std::reverse_iterator{context()->module()->types_values_end()},
+           itEnd =
+               std::reverse_iterator{context()->module()->types_values_begin()};
+       it != itEnd; ++it) {
```

Comment:
Is there a reason you do not use the def-use manager? If this is run as part of a larger set of optimizations, it will generally already be built. Then you can find the instruction for the type with a hash table look up.

Also, why do you do this in reverse? Is the struct type usually near the end?

Reply:
It's been a while since I originally wrote this, but while I reduced the `StructMemberInfo` struct to a simple `analysis::Type` field, I also replaced this code with just using `get_def_use_mgr()->GetDef()` to retrieve the respective struct instruction.